phoenix-core/src/main/java/org/apache/phoenix/compile/QueryCompiler.java

```diff
@@ -147,21 +169,146 @@ public QueryCompiler(PhoenixStatement statement, SelectStatement select, ColumnR
+
+    private QueryPlan buildTupleProjectPlan(QueryPlan plan) throws SQLException {
+        List<ExpressionProjector> projectedColumns = new ArrayList<ExpressionProjector>();
+        PTable tbl = createTempTableForUnionAllResultResolver(plan);
+        for (int i=0; i<tbl.getColumns().size(); i++) {
+            ProjectedColumnExpression expression = new ProjectedColumnExpression(tbl.getColumns().get(i), tbl.getColumns(), i, tbl.getColumns().get(i).getExpressionStr());
+            projectedColumns.add(new ExpressionProjector(tbl.getColumns().get(i).getName().getString(), tbl.getName().getString(), expression, true));
+        }
+        RowProjector rowProjector = new RowProjector(projectedColumns, 100, true);
+        TupleProjector tupleProjector = new TupleProjector(rowProjector);
+        plan = new TupleProjectionPlan(plan, tupleProjector, null);
+        return plan;
+    }
+    private boolean containOrderBy(SelectStatement select) {
+        if (select.getOrderBy() != null && !select.getOrderBy().isEmpty()) 
+            return true;
+        else
+            return false;
+    }
+
+    private QueryPlan compileUnionAll(SelectStatement select) throws SQLException { 
+        List<SelectStatement> unionAllSelects = select.getSelects();
+        List<QueryPlan> plans = new ArrayList<QueryPlan>();
         StatementContext context = new StatementContext(statement, resolver, scan, sequenceManager);
-        if (select.isJoin()) {
-            select = JoinCompiler.optimize(statement, select, resolver);
-            if (this.select != select) {
-                ColumnResolver resolver = FromCompiler.getResolverForQuery(select, statement.getConnection());
-                context = new StatementContext(statement, resolver, scan, sequenceManager);
+
+        checkForOrderByLimitInUnionAllSelect(select);
+        QueryPlan plan = compileSelect(select);
+        plan = buildTupleProjectPlan(plan);
+        plans.add(plan);
+        OrderBy orderBy = OrderBy.EMPTY_ORDER_BY;
+        boolean containOrderBy = false;
+        int numSelects = unionAllSelects.size();
+        for (int i=0; i < numSelects; i++ ) {
+            if (i < numSelects-1)
+                checkForOrderByLimitInUnionAllSelect(unionAllSelects.get(i));
+            else if (i == numSelects-1) {
+                containOrderBy = containOrderBy(unionAllSelects.get(i));
             }
-            JoinTable joinTable = JoinCompiler.compile(statement, select, context.getResolver());
-            return compileJoinQuery(context, binds, joinTable, false, false, null);
-        } else {
-            return compileSingleQuery(context, select, binds, false, true);
+            plan = compileSelect(unionAllSelects.get(i));
+            if (containOrderBy) {
+                orderBy = plan.getOrderBy();
+                unionAllSelects.get(i).removeOrderBy();
+                plan = compileSelect(unionAllSelects.get(i));
+            } 
+            plan = buildTupleProjectPlan(plan);
```

Comment:
Actually the buildTupleProjectPlan() is just a couple lines, maybe you don't even have to write a function for this. The code would like:

TupleProjector projector = new TupleProjector(plan.getRowProjector());
plan = new TupleProjectionPlan(plan, projector, null);